mapiproxy/libmapistore/mapistore_notification.c

```diff
@@ -3,525 +3,1355 @@
+	MAPISTORE_RETVAL_IF(retval, retval, mem_ctx);
+
+	rc = memcached_exist(mstore_ctx->notification_ctx->memc_ctx, key, strlen(key));
+	talloc_free(key);
+	MAPISTORE_RETVAL_IF(rc != MEMCACHED_SUCCESS, ret_to_mapistore(rc), mem_ctx);
+
+	talloc_free(mem_ctx);
+	return MAPISTORE_SUCCESS;
+}
+
+
+/**
+   \details Retrieve EMSMDB session UUID and user name associated to
+   specified asynchronous emsmdb session UUID
+
+   \param mem_ctx pointer to the memory context to use to allocate data to return
+   \param mstore_ctx pointer to the mapistore context
+   \param async_uuid the GUID of the asynchronous emsmdb session
+   \param uuidp pointer to the GUID of the emsmdb session
+   \param cnp pointer on pointer to the common name to return
+
+   \note calling function is responsible for freeing cnp
+
+   \return MAPISTORE_SUCCESS on success, otherwise MAPISTORE error
+ */
+_PUBLIC_ enum mapistore_error mapistore_notification_session_get(TALLOC_CTX *mem_ctx,
+								 struct mapistore_context *mstore_ctx,
+								 struct GUID async_uuid,
+								 struct GUID *uuid,
+								 char **cnp)
 {
-#if 0
-        struct mapistore_notification *new_notification;
-        struct mapistore_notification_list *new_list;
-        struct mapistore_table_notification_parameters *table_parameters;
-        struct mapistore_object_notification_parameters *object_parameters;
-
-        if (!mstore_ctx) return;
-
-	new_list = talloc_zero(mstore_ctx, struct mapistore_notification_list);
-	new_notification = talloc_zero(new_list, struct mapistore_notification);
-	new_list->notification = new_notification;
-	new_notification->object_type = object_type;
-	new_notification->event = event;
-	if (object_type == MAPISTORE_TABLE) {
-		table_parameters = parameters;
-		new_notification->parameters.table_parameters = *table_parameters;
+	TALLOC_CTX				*local_mem_ctx;
+	enum mapistore_error			retval;
+	enum ndr_err_code			ndr_err_code;
+	struct ndr_pull				*ndr;
+	struct mapistore_notification_session	r;
+	DATA_BLOB				blob;
+	char					*key = NULL;
+	char					*value = NULL;
+	size_t					value_len = 0;
+	memcached_return_t			rc;
+	uint32_t				flags;
+
+	/* Sanity checks */
+	MAPISTORE_RETVAL_IF(!mstore_ctx, MAPISTORE_ERR_NOT_INITIALIZED, NULL);
+	MAPISTORE_RETVAL_IF(!uuid, MAPISTORE_ERR_INVALID_PARAMETER, NULL);
+	MAPISTORE_RETVAL_IF(!cnp, MAPISTORE_ERR_INVALID_PARAMETER, NULL);
+	MAPISTORE_RETVAL_IF(!mstore_ctx->notification_ctx, MAPISTORE_ERR_NOT_AVAILABLE, NULL);
+	MAPISTORE_RETVAL_IF(!mstore_ctx->notification_ctx->memc_ctx, MAPISTORE_ERR_NOT_AVAILABLE, NULL);
+
+	local_mem_ctx = talloc_new(NULL);
+	MAPISTORE_RETVAL_IF(!local_mem_ctx, MAPISTORE_ERR_NO_MEMORY, NULL);
+
+	/* Retrieve key/value */
+	retval = mapistore_notification_session_set_key(local_mem_ctx, async_uuid, &key);
+	MAPISTORE_RETVAL_IF(retval, retval, local_mem_ctx);
+
+	value = memcached_get(mstore_ctx->notification_ctx->memc_ctx, key, strlen(key), &value_len,
+			      &flags, &rc);
+	talloc_free(key);
```

Comment:
I think this is not really needed because we will free `local_mem_ctx` in any case